src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -514,6 +556,111 @@ public class OpenIdConnectAuthenticationHandler : AuthenticationHandler<OpenIdCo
         }
 
         /// <summary>
+        /// Redeems the authorization code for tokens at the token endpoint
+        /// </summary>
+        /// <param name="authorizationCode">The authorization code to redeem.</param>
+        /// <param name="redirectUri">Uri that was passed in the request sent for the authorization code.</param>
+        /// <returns>OpenIdConnect message that has tokens inside it.</returns>
+        protected virtual async Task<OpenIdConnectMessage> RedeemAuthorizationCode(string authorizationCode, string redirectUri)
+        {
+            var openIdMessage = new OpenIdConnectMessage()
+            {
+                ClientId = Options.ClientId,
+                ClientSecret = Options.ClientSecret,
+                Code = authorizationCode,
+                GrantType = "authorization_code",
+                RedirectUri = redirectUri
+            };
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
+            requestMessage.Content = new FormUrlEncodedContent(openIdMessage.Parameters);
+            var responseMessage = await Backchannel.SendAsync(requestMessage);
+            responseMessage.EnsureSuccessStatusCode();
+            var tokenResonse = await responseMessage.Content.ReadAsStringAsync();
+            var jsonTokenResponse = JObject.Parse(tokenResonse);
```

Comment:
In OAuth we found that people wanted direct access to the jsonTokenResponse because sometimes there are additional parameters returned. Will this ever be true for OIDC? If so, should there be a notification at this point?